File: zipfend/services/sentry.ts
/**
 * Privacy-safe Sentry Error Monitoring Integration for ZipRIGHT Frontend.
 *
 * Features:
 * - Error monitoring only (tracing, profiling, and session replays disabled).
 * - Comprehensive privacy filtering:
 *   * Firebase auth tokens, session tokens, and credentials scrubbed.
 *   * Customer biometric data, avatar images, and try-on photos redacted.
 *   * Customer body measurements (fit profiles, SmartFit scans, bust/waist/hips) redacted.
 *   * Payment details and Razorpay signatures redacted.
 *   * Private chat and message content redacted.
 *   * User PII (email, phone, name) stripped; user ID non-reversibly hashed.
 *   * Sensitive breadcrumbs and query parameters sanitized.
 */

import * as Sentry from '@sentry/react';

// Regex patterns for sensitive keys across event contexts, extras, and payloads
const SENSITIVE_KEY_PATTERNS = [
  /(password|token|secret|credential|api_?key)/i,
  /(auth|authorization|bearer)/i,
  /(cookie|set-cookie)/i,
  /(razorpay|card|cvv|account_number)/i,
  /(x-wallet-topup-secret|x-razorpay-signature|x-csrf-token)/i,
  /(firebase_?credentials|id_?token|refresh_?token)/i,
  /(measurement|body_?shape|fit_?profile|smart_?fit|shoulder)/i,
  /^(bust|waist|hips|chest|height|weight|inseam|thigh)$/i,
  /(biometric|face_?crop|avatar_?image|selfie|photo_?url)/i,
  /(private_?media|signed_?url|download_?token)/i,
  /^(email|phone|phone_number|full_name|address|shipping_address)$/i,
  /^(message_text|chat_content|private_message)$/i,
];

const SENSITIVE_QUERY_PARAMS = new Set([
  'token',
  'secret',
  'key',
  'api_key',
  'password',
  'auth',
  'signature',
  'sig',
  'code',
  'access_token',
  'refresh_token',
]);

/**
 * Return true if key matches any known sensitive field pattern.
 */
export function isSensitiveKey(key: string): boolean {
  return SENSITIVE_KEY_PATTERNS.some((pattern) => pattern.test(key));
}

/**
 * Non-reversible simple hash for user IDs in browser context.
 */
export function hashUserId(userId?: string | number | null): string | undefined {
  if (!userId) return undefined;
  const str = String(userId).trim();
  if (!str) return undefined;

  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash |= 0; // Convert to 32bit integer
  }
  return `anon_${Math.abs(hash).toString(16)}`;
}

/**
 * Recursively sanitize objects and arrays to redact sensitive keys and values.
 */
export function sanitizeData<T>(data: T): T {
  if (data === null || data === undefined) {
    return data;
  }

  if (typeof data === 'string') {
    // Redact token in Firebase Storage or bearer links
    if (data.includes('firebasestorage.googleapis.com') && data.includes('token=')) {
      return data.replace(/token=[^&\s]+/g, 'token=[REDACTED]') as unknown as T;
    }
    if (data.includes('Bearer ')) {
      return data.replace(/Bearer\s+[^\s]+/g, 'Bearer [REDACTED]') as unknown as T;
    }
    return data;
  }

  if (Array.isArray(data)) {
    return data.map((item) => sanitizeData(item)) as unknown as T;
  }

  if (typeof data === 'object') {
    const sanitized: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
      if (isSensitiveKey(key)) {
        sanitized[key] = '[REDACTED]';
      } else {
        sanitized[key] = sanitizeData(value);
      }
    }
    return sanitized as unknown as T;
  }

  return data;
}

/**
 * Redact sensitive query parameters in a URL string.
 */
export function sanitizeUrl(url?: string): string {
  if (!url) return '';
  const [base, query] = url.split('?');
  if (!query) return base;

  const cleanParams = query.split('&').map((param) => {
    const [k, v] = param.split('=');
    if (k && (SENSITIVE_QUERY_PARAMS.has(k.toLowerCase()) || isSensitiveKey(k))) {
      return `${k}=[REDACTED]`;
    }
    return param;
  });

  return `${base}?${cleanParams.join('&')}`;
}

/**
 * Central event scrubber applied before sending to Sentry.
 */
export function beforeSendFilter(
  event: Sentry.ErrorEvent,
  _hint?: Sentry.EventHint
): Sentry.ErrorEvent | null {
  try {
    // 1. Scrub User Information
    if (event.user) {
      delete event.user.email;
      delete event.user.username;
      delete event.user.ip_address;
      // Strip any extra PII attached to user
      const customUser = event.user as Record<string, unknown>;
      delete customUser.phone;
      delete customUser.name;

      if (event.user.id) {
        event.user.id = hashUserId(event.user.id);
      }
    }

    // 2. Scrub Request URL & Query String
    if (event.request?.url) {
      event.request.url = sanitizeUrl(event.request.url);
    }
    if (event.request?.headers) {
      event.request.headers = sanitizeData(event.request.headers);
    }
    if (event.request?.data) {
      event.request.data = sanitizeData(event.request.data);
    }

    // 3. Scrub Contexts
    if (event.contexts) {
      event.contexts = sanitizeData(event.contexts);
    }

    // 4. Scrub Extra Data
    if (event.extra) {
      event.extra = sanitizeData(event.extra);
    }

    // 5. Scrub Breadcrumbs
    if (event.breadcrumbs) {
      event.breadcrumbs = event.breadcrumbs.map((crumb) => beforeBreadcrumbFilter(crumb));
    }
  } catch (err) {
    // Fallback: don't block error dispatch if scrubbing encounters unexpected structure
    console.warn('Error during Sentry beforeSend filtering:', err);
  }

  return event;
}

/**
 * Breadcrumb scrubber to prevent capturing sensitive network/navigation data.
 */
export function beforeBreadcrumbFilter(
  breadcrumb: Sentry.Breadcrumb,
  _hint?: Sentry.BreadcrumbHint
): Sentry.Breadcrumb {
  if (breadcrumb.data) {
    const data = breadcrumb.data as Record<string, unknown>;
    if (typeof data.url === 'string') {
      data.url = sanitizeUrl(data.url);
    }
    breadcrumb.data = sanitizeData(data);
  }

  if (breadcrumb.message) {
    if (breadcrumb.message.includes('firebasestorage.googleapis.com') && breadcrumb.message.includes('token=')) {
      breadcrumb.message = breadcrumb.message.replace(/token=[^&\s]+/g, 'token=[REDACTED]');
    }
    if (breadcrumb.message.includes('Bearer ')) {
      breadcrumb.message = breadcrumb.message.replace(/Bearer\s+[^\s]+/g, 'Bearer [REDACTED]');
    }
  }

  return breadcrumb;
}

let sentryInitialized = false;

export function isSentryInitialized(): boolean {
  return sentryInitialized;
}

function getEnvVar(name: string): string | undefined {
  try {
    if (typeof import.meta !== 'undefined' && import.meta.env && typeof import.meta.env[name] === 'string') {
      return import.meta.env[name];
    }
  } catch {
    // import.meta may not be available in node test runner
  }
  if (typeof process !== 'undefined' && process.env && typeof process.env[name] === 'string') {
    return process.env[name];
  }
  return undefined;
}

/**
 * Initialize Sentry for React/Vite.
 * Only activates if VITE_SENTRY_DSN is configured.
 */
export function initSentry(): boolean {
  const dsn = getEnvVar('VITE_SENTRY_DSN');
  if (!dsn || !dsn.trim()) {
    return false;
  }

  const environment =
    getEnvVar('VITE_SENTRY_ENVIRONMENT') || getEnvVar('MODE') || 'development';
  const release = getEnvVar('VITE_SENTRY_RELEASE') || 'zipright-frontend@1.0.0';

  Sentry.init({
    dsn: dsn.trim(),
    environment,
    release,
    // Phase 5B Quota & Privacy:
    sendDefaultPii: false,
    tracesSampleRate: 0,           // Tracing DISABLED
    replaysSessionSampleRate: 0,   // Session Replay DISABLED
    replaysOnErrorSampleRate: 0,   // Replay on Error DISABLED
    maxBreadcrumbs: 30,
    beforeSend: beforeSendFilter,
    beforeBreadcrumb: beforeBreadcrumbFilter,
  });

  sentryInitialized = true;
  return true;
}

File: zipfin-backend/core/sentry.py
"""Privacy-safe Sentry integration for ZipRIGHT backend.

Features:
- Error monitoring only (traces, profiling, replays, metrics explicitly disabled).
- Comprehensive privacy filter:
  * Sensitive headers scrubbed (Authorization, Cookie, Secrets, Signatures, Keys).
  * Sensitive request data / body fields redacted (passwords, payment credentials,
    biometrics, fit/body measurements, private media URLs, tokens).
  * User PII stripped (email, phone, name, ip_address).
  * User IDs non-reversibly hashed with SHA-256.
  * Breadcrumbs scrubbed of sensitive query parameters and auth headers.
"""

from __future__ import annotations

import hashlib
import logging
import re
from typing import Any

from core.config import settings

logger = logging.getLogger(__name__)

# Headers that must be completely redacted from Sentry events
SENSITIVE_HEADER_PATTERNS = [
    re.compile(r"^authorization$", re.I),
    re.compile(r"^cookie$", re.I),
    re.compile(r"^set-cookie$", re.I),
    re.compile(r"^x-wallet-topup-secret$", re.I),
    re.compile(r"^x-razorpay-signature$", re.I),
    re.compile(r"^x-csrf-token$", re.I),
    re.compile(r"^x-api-key$", re.I),
    re.compile(r".*(token|secret|key|auth|credential|password|signature).*", re.I),
]

# Field names across request data, contexts, and extras that must be redacted
SENSITIVE_FIELD_PATTERNS = [
    re.compile(r".*(password|token|secret|credential|api_?key).*", re.I),
    re.compile(r".*(razorpay|card|cvv|account_number).*", re.I),
    re.compile(r".*(firebase_credentials|service_account|private_key).*", re.I),
    re.compile(r".*(measurement|body_?shape|fit_?profile|smart_?fit|shoulder).*", re.I),
    re.compile(r"^(bust|waist|hips|chest|height|weight|inseam|thigh)$", re.I),
    re.compile(r".*(biometric|face_?crop|avatar_?image|selfie).*", re.I),
    re.compile(r".*(private_media|signed_url|download_token).*", re.I),
    re.compile(r"^(email|phone|phone_number|full_name|shipping_address|address)$", re.I),
    re.compile(r"^(message_text|chat_content|private_message)$", re.I),
]

# URL query parameter keys that must be redacted
SENSITIVE_QUERY_PARAMS = {
    "token", "secret", "key", "api_key", "password", "auth",
    "signature", "sig", "code", "access_token", "refresh_token",
}


def hash_user_id(user_id: str | int | None) -> str | None:
    """Return a non-reversible truncated SHA-256 hash of the user ID."""
    if not user_id:
        return None
    raw = str(user_id).strip()
    if not raw:
        return None
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:16]


def is_sensitive_key(key: str) -> bool:
    """Check if a dictionary key matches any sensitive field pattern."""
    return any(pattern.search(key) for pattern in SENSITIVE_FIELD_PATTERNS)


def sanitize_value(key: str, value: Any) -> Any:
    """Recursively scrub sensitive keys and nested dictionaries/lists."""
    if is_sensitive_key(key):
        return "[REDACTED]"

    if isinstance(value, dict):
        return {k: sanitize_value(k, v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [sanitize_value(key, item) for item in value]
    if isinstance(value, str):
        # Scrub potential private media or bearer tokens embedded in strings
        if "firebasestorage.googleapis.com" in value and "token=" in value:
            return re.sub(r"token=[^&\s]+", "token=[REDACTED]", value)
        if "Bearer " in value:
            return re.sub(r"Bearer\s+[^\s]+", "Bearer [REDACTED]", value)
    return value


def sanitize_headers(headers: dict[str, Any]) -> dict[str, Any]:
    """Scrub sensitive request headers."""
    sanitized = {}
    for header, value in headers.items():
        if any(pattern.match(header) for pattern in SENSITIVE_HEADER_PATTERNS):
            sanitized[header] = "[REDACTED]"
        else:
            sanitized[header] = value
    return sanitized


def sanitize_query_string(query_string: str) -> str:
    """Redact sensitive query parameter values."""
    if not query_string:
        return ""
    parts = query_string.split("&")
    clean_parts = []
    for part in parts:
        if "=" in part:
            k, v = part.split("=", 1)
            if k.lower() in SENSITIVE_QUERY_PARAMS or is_sensitive_key(k):
                clean_parts.append(f"{k}=[REDACTED]")
            else:
                clean_parts.append(part)
        else:
            clean_parts.append(part)
    return "&".join(clean_parts)


def sanitize_url(url: str) -> str:
    """Redact sensitive query parameters in a full URL."""
    if not url:
        return ""
    if "?" in url:
        base, query = url.split("?", 1)
        return f"{base}?{sanitize_query_string(query)}"
    return url


def before_send(event: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any] | None:
    """Central event scrubber applied immediately before an event is dispatched to Sentry."""
    try:
        # 1. Scrub Request Information
        req = event.get("request")
        if isinstance(req, dict):
            # Headers
            if "headers" in req and isinstance(req["headers"], dict):
                req["headers"] = sanitize_headers(req["headers"])

            # Cookies
            if "cookies" in req:
                req["cookies"] = "[REDACTED]"

            # Query string
            if "query_string" in req and isinstance(req["query_string"], str):
                req["query_string"] = sanitize_query_string(req["query_string"])

            # URL
            if "url" in req and isinstance(req["url"], str):
                req["url"] = sanitize_url(req["url"])

            # Request body / data
            if "data" in req:
                req["data"] = sanitize_value("data", req["data"])

        # 2. Scrub User Information
        user = event.get("user")
        if isinstance(user, dict):
            # Strip all direct PII
            user.pop("email", None)
            user.pop("phone", None)
            user.pop("username", None)
            user.pop("name", None)
            user.pop("ip_address", None)

            # Hash user ID non-reversibly
            raw_id = user.get("id")
            if raw_id:
                user["id"] = hash_user_id(raw_id)
            else:
                user.pop("id", None)

        # 3. Scrub Contexts & Extras
        for container_key in ("contexts", "extra"):
            container = event.get(container_key)
            if isinstance(container, dict):
                event[container_key] = sanitize_value(container_key, container)

        # 4. Scrub Breadcrumbs attached to the event
        crumbs = event.get("breadcrumbs")
        if isinstance(crumbs, dict) and "values" in crumbs:
            for crumb in crumbs["values"]:
                _sanitize_single_breadcrumb(crumb)
        elif isinstance(crumbs, list):
            for crumb in crumbs:
                _sanitize_single_breadcrumb(crumb)

    except Exception as exc:
        logger.warning("Error during Sentry before_send sanitization: %s", exc)

    return event


def _sanitize_single_breadcrumb(crumb: dict[str, Any]) -> None:
    """In-place sanitizer for a single breadcrumb."""
    if not isinstance(crumb, dict):
        return

    # Sanitize URL data in HTTP breadcrumbs
    data = crumb.get("data")
    if isinstance(data, dict):
        if "url" in data and isinstance(data["url"], str):
            data["url"] = sanitize_url(data["url"])
        crumb["data"] = sanitize_value("data", data)

    # Sanitize breadcrumb message
    msg = crumb.get("message")
    if isinstance(msg, str):
        if "firebasestorage.googleapis.com" in msg and "token=" in msg:
            crumb["message"] = re.sub(r"token=[^&\s]+", "token=[REDACTED]", msg)


def before_breadcrumb(crumb: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any] | None:
    """Filter and sanitize breadcrumbs before recording."""
    try:
        _sanitize_single_breadcrumb(crumb)
    except Exception as exc:
        logger.warning("Error during Sentry before_breadcrumb sanitization: %s", exc)
    return crumb


def init_sentry() -> bool:
    """Initialize Sentry with strict privacy safeguards.

    Returns True if initialized, False if skipped (e.g. SENTRY_DSN not configured).
    """
    dsn = settings.SENTRY_DSN
    if not dsn:
        logger.info("SENTRY_DSN is not configured; Sentry monitoring is disabled.")
        return False

    import sentry_sdk
    from sentry_sdk.integrations.fastapi import FastApiIntegration
    from sentry_sdk.integrations.starlette import StarletteIntegration

    environment = settings.SENTRY_ENVIRONMENT
    release = settings.SENTRY_RELEASE

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        # Phase 5B Quota & Privacy settings:
        send_default_pii=False,
        traces_sample_rate=0.0,      # Tracing DISABLED
        profiles_sample_rate=0.0,    # Profiling DISABLED
        enable_tracing=False,
        attach_stacktrace=True,
        max_breadcrumbs=30,
        before_send=before_send,
        before_breadcrumb=before_breadcrumb,
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            StarletteIntegration(transaction_style="endpoint"),
        ],
    )
    logger.info("Sentry initialized successfully for environment '%s'.", environment)
    return True
